cpp/models/abm/world.cpp

```diff
@@ -38,20 +34,37 @@ namespace abm
 LocationId World::add_location(LocationType type, uint32_t num_cells)
 {
     LocationId id = {static_cast<uint32_t>(m_locations.size()), type};
-    m_locations.emplace_back(std::make_unique<Location>(id, parameters.get_num_groups(), num_cells));
+    m_locations.emplace_back(id, parameters.get_num_groups(), num_cells);
     m_has_locations[size_t(type)] = true;
+
+    // mark caches for rebuild
+    m_is_local_population_cache_valid = false;
+    m_are_exposure_caches_valid       = false;
+    m_exposure_caches_need_rebuild    = true;
+
     return id;
 }
 
-Person& World::add_person(const LocationId id, AgeGroup age)
+PersonId World::add_person(const LocationId id, AgeGroup age)
+{
+    return add_person(Person(m_rng, id, age));
+}
+
+PersonId World::add_person(Person&& person)
 {
-    assert(age.get() < parameters.get_num_groups());
-    uint32_t person_id = static_cast<uint32_t>(m_persons.size());
-    m_persons.push_back(std::make_unique<Person>(m_rng, get_individualized_location(id), age, person_id));
-    auto& person = *m_persons.back();
-    person.set_assigned_location(m_cemetery_id);
-    get_individualized_location(id).add_person(person);
-    return person;
+    assert(person.get_location().index != INVALID_LOCATION_INDEX);
+    assert(person.get_location().index < m_locations.size());
+    assert(person.get_age().get() < parameters.get_num_groups());
+
+    PersonId new_id{static_cast<uint32_t>(m_persons.size())};
+    m_persons.emplace_back(person, new_id);
+    auto& new_person = m_persons.back();
+    new_person.set_assigned_location(m_cemetery_id);
+
+    if (m_is_local_population_cache_valid) {
+        ++m_local_population_cache[new_person.get_location().index];
+    }
```

Comment:
Please add a test for `m_is_local_population_cache_valid`

Reply:
There is no meaningful way to test this line, as it is entirely internal model state. It rarely makes sense to test internal states, and this line's effect is basically impossible to observe, as is just a tiny optimization for when persons (but no locations!) get added after a begin_step has taken place.

I would honestly rather remove it and just invalidate the cache instead of writing a test just for this, since for now this optimization is completely unused. Or we accept that this is left uncovered.

Reply:
Yes, I can see it's not covered also in the main branch. I guess we can leave it for now.